plugins/inputs/tacacs/tacacs.go

```diff
@@ -0,0 +1,164 @@
+	// Create the fields for this metric
+	tags := map[string]string{"source": server}
+	fields := make(map[string]interface{})
+
+	secret, err := t.Secret.Get()
+	if err != nil {
+		return fmt.Errorf("getting secret failed: %v", err)
+	}
+	defer config.ReleaseSecret(secret)
+
+	username, err := t.Username.Get()
+	if err != nil {
+		return fmt.Errorf("getting username failed: %v", err)
+	}
+	defer config.ReleaseSecret(username)
+
+	password, err := t.Password.Get()
+	if err != nil {
+		return fmt.Errorf("getting password failed: %v", err)
+	}
+	defer config.ReleaseSecret(password)
+
+	// Create the tacacs client
+	client := &tacplus.Client{
+		Addr: server,
+		ConnConfig: tacplus.ConnConfig{
+			Secret: secret,
+		},
+	}
+
+	// Initialize the AuthStart data
+	testAuthStart := &tacplus.AuthenStart{
+		Action:        tacplus.AuthenActionLogin,
+		AuthenType:    tacplus.AuthenTypeASCII,
+		AuthenService: tacplus.AuthenServiceLogin,
+		PrivLvl:       1,
+		Port:          "heartbeat",
+		RemAddr:       t.RemAddr,
+	}
+
+	// send the start request, the reply should be AuthenStatusGetUser
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(t.ResponseTimeout))
+	defer cancel()
+	startTime := time.Now()
+	reply, session, err := client.SendAuthenStart(ctx, testAuthStart)
+	if err != nil {
+		t.Log.Warnf("error on new tacacs authentication start request to %s : %s", server, err)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
+	defer session.Close()
+	// Check the returned status
+	if reply.Status != tacplus.AuthenStatusGetUser {
+		t.Log.Warnf("error on new tacacs authentication start request to %s : Unexpected response code %d", server, reply.Status)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
+
+	// Send the first Continue request with the username, the reply should be AuthenStatusGetPass
+	reply, err = session.Continue(ctx, string(username))
+	if err != nil {
+		t.Log.Warnf("error on new tacacs authentication continue username request to %s : %s", server, err)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
+	if reply.Status != tacplus.AuthenStatusGetPass {
+		t.Log.Warnf("error on first tacacs continue username request to %s : Unexpected response code %d", server, reply.Status)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
```

Comment:
What about adding status in the returned metric?

Reply:
> Tried using the status code returned, but it was very useless as the status depends on the current tacacs auth sequence. On initial start packet, return code X may mean something completely different than return code X for send username packet, then X may mean something else on send password packet etc. So without the secondary info at what stage the status code was returned, its useless. In the radius plugin, the status code is always from the same place - its just one request to do the auth, not multiple, so its nicely descriptive with exact string representation of the status. Here its not the case, you get an integer code that may mean something else for every auth stage.
> 
> Anyway, its meant to monitor SUCCESSFUL authentications response time, so decided to simply return an error in all other cases. Let me know if that makes sense ?

As far as I can see, the status code would only be one of these values?
https://github.com/nwaples/tacplus/blob/v0.0.3/packet.go#L62-L72